Support for mmap() writable mappings.

aversecat opened the pull request:
Replaces #27, #39.

Contains ~~mostly original patches from andy, touched up for conflicts~~. Additional fixups and changes to avoid various deadlocks and debug kernel warnings for lock contention issues.

- [x] - ~~Does not~~ pass `xfstests:generic/346` - hard lockup in _mkwrite when doing update_inode
- [x] - ~~Occasionally fails~~ offline-extent-waiting - when reverse staging, the first blocks of the file end up zeros, not the expected content 
- [x] - Passes all other xfstests 
- [x] - Added cross-node mmap consistency test.  ~~doesn't work on `el7`~~
- [x] - sparse warnings about ret not returning vm_fault_t
- [x] - _walk_inodes page fault safe
- [x] - _get_allocated_inos page fault safe
- [x] - fsstress hard lockup in generic/013

versity-github commented:
Build 169 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/169/ git: c6a7fd96ab4b475c018c49e46b8f761ef752a231

versity-github commented:
Build 144 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/144/ git: c6a7fd96ab4b475c018c49e46b8f761ef752a231

versity-github commented:
Build 170 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/170/ git: 0f611dfbf8826f70fb25bff980c26c8eb757f7d8

versity-github commented:
Build 145 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/145/ git: 0f611dfbf8826f70fb25bff980c26c8eb757f7d8

versity-github commented:
Build 178 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/178/ git: d112efbf064d4ca520b6bcd1ba94487ac560bc5c

versity-github commented:
Build 153 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/153/ git: d112efbf064d4ca520b6bcd1ba94487ac560bc5c

versity-github commented:
Build 180 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/180/ git: 43de8172fbde647d0264e6e33b6e87793ecdb650

versity-github commented:
Build 155 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/155/ git: 43de8172fbde647d0264e6e33b6e87793ecdb650

versity-github commented:
Build 156 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/156/ git: 939b7ec1e20dce3363da625eb8db541e4221bb36

versity-github commented:
Build 182 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/182/ git: 939b7ec1e20dce3363da625eb8db541e4221bb36

versity-github commented:
Build 185 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/185/ git: cb51a815c27f7ae49e9595b1749608402464f2b1

versity-github commented:
Build 160 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/160/ git: cb51a815c27f7ae49e9595b1749608402464f2b1

zabbo commented on tests/tests/mmap.sh:
```diff
@@ -0,0 +1,186 @@
+# test mmap() and normal read/write consistency between different nodes
+#
+# This script opens a file with a fixed size and writes a pattern
+# at random offset with random length, and then reads it again.
```
Ah.  I had pictured a simpler ping pong unit test to just make sure that the basics worked.  If we're getting into random (ish) fuzz testing then we should use fsx-mpi.  It's a bit like this, but it races the operations, can do more (truncate), and stops and gives you a dump of the problem and operations that lead to it when it sees a problem.

aversecat replied:
I've repaced it with a C program which is simpler and slightly faster than doing this using xfs_io and scripts, and as a bonus it just works on el7 as well. We can do fsx-mpi stuff maybe later, I'm thinking. For now this does a good job at hitting the lock contention.

zabbo commented on kmod/src/data.c:
```diff
@@ -1915,22 +1915,23 @@ static int scoutfs_data_page_mkwrite(struct vm_area_struct *vma,
 	int err;
 
 	sb_start_pagefault(sb);
+	inode_lock(inode);
```
I'd be surprised if this doesn't deadlock somewhere.  It certainly would in our code, but the bigger question is the rest of the kernel attempting to copy to userspace while holding the inode lock.  Either this needs to show why this is safe or, more likely, relax the inode lock requirement in get_block.  If it's just for serializing access to extents the extent_sem should do that.

zabbo commented on tests/golden/xfstests:
```diff
@@ -9,9 +9,9 @@ generic/011
-generic/024
 generic/028
+generic/029
+generic/030
```
I guess this would be coordinated with updating our xfstests?

aversecat replied:
Hmm I'll double check all these to make sure this isn't from an e9 branch....

aversecat replied:
I swuashed the patch that re-enables mmap() xfstests and the one that adds the new `golden` output and it all properly corresponds now. There's no new branch needed here - this is all el8/el7 tests from the `scoutfs` branch.

zabbo commented on kmod/src/data.c:
```diff
@@ -1889,6 +1889,127 @@ int scoutfs_data_waiting(struct super_block *sb, u64 ino, u64 iblock,
 	return ret;
 }
 
+/* Mostly cribbed from mm/filemap.c:filemap_page_mkwrite() */
```
Is what's cribbed still appropriate for the modern kernel paths?  Checking that would probably resolve the later patches that seem uncertain about whether or not they're doing it right.  See where the current vfs and file systems do the things this cribbed code, and those fixes, are doing and make ours do the same.

aversecat replied:
I removed this comment entirely.

In a way this comment makes sense since filemap_page_mkwrite() literally is a skeleton function here but there's not a single filesystem that actually uses it, it's only used by `aio_ring`.

It is however still appropriate because it shows the basic locking order properly. And this function hasn't changed a single line of code from 4.x all the way to current mainline, it's still the exact same function.

But it doesn't do any actual physical data updates ... for that reason it's value is somewhat bleak.

The only thing that will change is page vs. folio, and we can address that in the `el9` patch set.

zabbo reviewed (changes requested):
It's a WIP, we knew there'd be W to do :).

versity-github commented:
Build 163 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/163/ git: 3db7fd0d229c6c4039fd2bc106b317705c4360b0

versity-github commented:
Build 188 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/188/ git: 3db7fd0d229c6c4039fd2bc106b317705c4360b0

versity-github commented:
Build 189 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/189/ git: 920070dd187e08e42b37357f4dd9510621935ece

versity-github commented:
Build 165 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/165/ git: 920070dd187e08e42b37357f4dd9510621935ece

versity-github commented:
Build 190 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/190/ git: 789959f361ed0c45177d7adc464d7462a9032063

versity-github commented:
Build 166 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/166/ git: 789959f361ed0c45177d7adc464d7462a9032063

versity-github commented:
Build 167 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/167/ git: d958c43a2bb1060f315e0f28c0d939e6f4be0c53

aversecat commented:
Pushed 1 more commit that adds the 2 basic mmap() trace events.

versity-github commented:
Build 192 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/192/ git: 37bcb61835931f9637147c394d7441aba133ae6a

versity-github commented:
Build 168 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/168/ git: 37bcb61835931f9637147c394d7441aba133ae6a

zabbo commented on kmod/src/data.c:
```diff
@@ -1583,30 +1587,40 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 
+	for (pos = 0; pos < SCOUTFS_DATA_FIEMAP_MAX; pos++) {
+		if (cur[pos].len) {
+			ret = fill_extent(fieinfo, &cur[pos], pos == (last_pos - 1) ? FIEMAP_EXTENT_LAST : 0);
```
This loop rewrite has a bunch of issues.  It never merges 'cause it always advances pos after storing an extent in cur.  It can lose the _LAST flag if it hits enoent after emptying cur.  It loses the ret = 1 -> 0 translation.  I think it can loop for ever if iblock runs into last without setting last_pos or ret?  The different types of pos and last_pos is confusing.  At a high level, the fiemap interface is so annoying because we have to read the next extent before returning an extent so that we can find out if we have to set _LAST.  I think what'd be clearer is having a main outer loop like the current iblock <= last, then two inner loops.  One that merges and prepares extents with _next and the other which drains the prepared array by calling the kernel's fiemap_fill_next_extent().  It could use the kernel's fiemap_info struct for the allocated array storage so that the _next loop could set the LAST flag without needing the variable state (last_flags/last_pos) between the two loops.  But critically, the second loop would leave one extent remaining so that the next _next call can see enoent and set the _LAST flag on that remaining extent.

aversecat replied:
I've completely rewritten this and the readdir one as well to be in line with these comments.

zabbo commented on kmod/src/dir.c:
```diff
@@ -449,67 +450,114 @@ static int scoutfs_readdir(struct file *file, struct dir_context *ctx)
 		init_dirent_key(&key, SCOUTFS_READDIR_TYPE, scoutfs_ino(inode),
 				ctx->pos, 0);
 
+		dent = (struct scoutfs_dirent *)(data + page_pos + sizeof(int));
```
The main problem with all this is unaligned accesses.  dirent_bytes() can be byte aligned, so the name_len and dent pointers can end up unaligned.  That's fine on x86, but very not fine on everything else.  (And may break on debug tests depending on how aggressively unforgiving they are.)  I think the fix here is a lot like fiemap.  An outer loop with two inner loops to get rid of the gotos.  And you can just iterate through dent pointers if you have a dirent_bytes() variant that aligns.  Have an end that's page_address(page) + PAGE_SIZE.  Then it can fill while next_aligned_dirent(dent, SCOUTFS_NAME_LEN) < end.  After reading each you can have dent = next_aligned_dirent(dent, len).  And you can store name_len in one of the dirent _pad fields to avoid storing it separately.  Finally, it needs to be more careful about setting ctx->pos.  The caller will unconditionally set the file->pos to ctx->pos, so it should really only be set after successfully emiting an entry.  The setting of it to the next entry about to be returned is a micro optimization that we could just leave out if it's too confusing.  And we don't need all the out_ exit labels.  Set the lock to NULL after you unlock it in the null and an _unlock in the exit path will be safe along with kfree() and the guarded page free so everything can always use the same exit path.

aversecat replied:
same here, we'll redo the review.

zabbo commented on kmod/src/kernelcompat.h:
```diff
@@ -279,4 +279,9 @@ ssize_t kc_generic_file_buffered_write(struct kiocb *iocb, const struct iovec *i
 
+#ifndef KC_MM_VM_FAULT_T
+typedef unsigned int vm_fault_t;
+#define vmf_error(err) err
```
We probably want (err) to avoid (admitedly unlikely) hard to debug surprises.

aversecat replied:
yep

zabbo reviewed (changes requested):
I still need to get to the core mmap ops patches.  I also did a quick check of other user access calls in our code and there's a few more to fix.  _walk_inodes and _get_allocated_inos both touch user mappings under cluster locking.  (And the comment above walk_inodes is ancient.. that may have been true once, but it can deadlock now).

versity-github commented:
Build 193 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/193/ git: a99e2a55cdbc920582bb3102f946d04caadc159a

versity-github commented:
Build 169 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/169/ git: a99e2a55cdbc920582bb3102f946d04caadc159a

aversecat commented on kmod/src/data.c:
```diff
@@ -1531,9 +1531,11 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 	struct scoutfs_extent ext;
-	struct scoutfs_extent cur;
+#define SCOUTFS_DATA_FIEMAP_MAX 32 /* 1k */
+	struct scoutfs_extent cur[SCOUTFS_DATA_FIEMAP_MAX];
```
Yeah, I had 4k there at first and GCC violently told me it wasn't cool with that. We'll grab a page for this.

aversecat commented on tests/src/mmap_stress.c:
```diff
@@ -0,0 +1,146 @@
+		/* write */
+		if (i == 0) {
+			for (written = 0; written < count;) {
+				ret = pwrite(fd_in, buf, count - written, written);
```
I just have it all swapped. The userspace process writes to fd_in, and reads from fd_out. Could rename it but it's just semantics.

aversecat replied:
This is now rewritten:

- spawns 5 threads, each accessing the same file from /mnt/test.{0,5}/
- randomly performs a pread, pwrite, mmap() read or write (seems evenly enough spread - about 5% variation across operations).
- size+count can be passed on cmdline

on my el8 debug kernel it hits about 20k operations in ~30 seconds. On non-debug it's ~3x faster. 

no consistency checking still. We could do that, but I'd have to make some sort of semaphore to sync all the threads so that they all know what read pattern to expect when verifying content. Probably worth doing.

aversecat commented:
I still need to get to _walk_inodes and _get_allocated_inos, but all the other commented parts have undergone significant rewrites and have been fully retested, so, we're a good few steps ahead where this was a week ago.

versity-github commented:
Build 200 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/200/ git: 417d19769eef137ccaa24b58279f0c40dda8288d

versity-github commented:
Build 176 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/176/ git: 417d19769eef137ccaa24b58279f0c40dda8288d

versity-github commented:
Build 201 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/201/ git: 983bb2f23db00c687f02919e7a6cce236758e903

versity-github commented:
Build 177 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/177/ git: 983bb2f23db00c687f02919e7a6cce236758e903

zabbo commented on kmod/src/data.c:
```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
+
+		ret = scoutfs_lock_inode(sb, SCOUTFS_LOCK_READ, 0, inode, &lock);
+		if (ret) {
+			lock = NULL;
```
This error path doesn't unlock the inode or release the extent_sem.  It also doesn't need to set the lock pointer to null here, the lock functions specifically set the lock pointer to null when they return an error.

aversecat replied:
Yeah, good catch.

zabbo commented on kmod/src/data.c:
```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
+			trace_scoutfs_data_fiemap_extent(sb, ino, &ext);
+
+			if (ext.start > last) {
+				/* not setting _LAST, it's for end of file */
```
This can now loop forever 'cause it's not breaking out of the outer iblock < last.  You'd need to set iblock here to break out (or have a complete like the other loops).  I guess the same's true of the ENOENT->LAST break as well (once we fix the later last_flags test and break).

aversecat replied:
I guess this is going to be easier to get consistent with a complete flag. At least to me.

zabbo commented on kmod/src/data.c:
```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
+				cur.len += ext.len;
+			} else {
+				/* fill it */
+				info->fe_logical = cur.start;
```
The idea behind using the kernel fiemap struct in the page was to be able to pass them directly to the kernel's fill extent.  If you're bouncing them back into the scoutfs_extent before calling fill then the page storage might as well just use the scoutfs extents.  (The dissonance that triggers this realization is that it looks weird to store the block granular start in the byte granular fe_logical.)

aversecat replied:
I'll stick it to scoutfs_extent for now, since it's the smaller of the options, resulting in less lock/unlock cycles.

zabbo commented on kmod/src/data.c:
```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
-		} else {
-			ret = fill_extent(fieinfo, &cur, 0);
+		/* unlock */
+		if (lock)
```
Don't need the lock pointer test here.

zabbo commented on kmod/src/data.c:
```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 
-		iblock = ext.start + ext.len;
+		/* still one left, it's in cur */
+		if (last_flags) {
```
Calling fill_extent on cur isn't limited to when we've set the last_flags.  If the fiemap region doesn't cover the end of the file then we'll want to emit the final extent that was being merged as we hit the last iblock of the range.  It needs to test cur.len in that case.  Once it's doing that you can't try that every iteration, so you move it out of the loop that's emitting the page extents.  (And now it won't break out of having set last_flags on ENOENT, so this is where we'd need to propogate out those exit conditions.)

zabbo commented on kmod/src/data.c:
```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
-	inode_unlock(inode);
-
 out:
-	if (ret == 1)
```
I'm not sure why this was removed.  It's here 'cause fill_extent can return 1 when it should stop filling and return and we didn't want to let that escape our inner loop.

aversecat replied:
Overlooked this and it'll be put back.

aversecat commented on kmod/src/ioctl.c:
```diff
@@ -1212,46 +1216,81 @@ static long scoutfs_ioc_get_allocated_inos(struct file *file, unsigned long arg)
 		goto out;
 	}
 
+	page = alloc_page(GFP_KERNEL);
```
Doh, yes.

zabbo commented on kmod/src/data.c:
```diff
@@ -1889,6 +1889,77 @@ int scoutfs_data_waiting(struct super_block *sb, u64 ino, u64 iblock,
+		inode_dio_done(inode);
+	scoutfs_unlock(sb, inode_lock, SCOUTFS_LOCK_READ);
+	if (scoutfs_data_wait_found(&dw)) {
+		int err = scoutfs_data_wait(inode, &dw);
```
There's a lot of confusing stuff around the way this calls data_wait (you can't have have_ret and _data_wait_found both be true.  But honestly, I'm not sure if it needs to do any of this at all.  filemap_fault will call ->readpage which will wait for the data.  So tear all the waiting stuff out from here and write a quick test to use xfs_io to do an mmap read from an offline file and see if it triggers staging and gets the correct data?

aversecat replied:
So, I believe that this code is very much still needed. I've modified the `tests/mmap.sh` script in the next revision to offline an extent and then test `mread` and `mwrite` on that using xfs_io, and with the data wait stuff in this function, every thing is just happy (Aside the have_ret confusion).

If I remove any of the data wait stuff, it all falls over and mread() returns all zeroes instead of the expected pattern - clearly there's something at hand here that this is still needed.

The next iteration of this code will have the `have_ret` stuff removed out of this particular code path because indeed that's not right the way it is, and that holds up nicely in testing.

aversecat commented on kmod/src/ioctl.c:
```diff
@@ -113,77 +115,109 @@ static long scoutfs_ioc_walk_inodes(struct file *file, unsigned long arg)
-		ret = scoutfs_item_next(sb, &key, &last_key, NULL, 0, lock);
-		if (ret < 0 && ret != -ENOENT)
+		ret = scoutfs_lock_inode_index(sb, SCOUTFS_LOCK_READ, type,
+					       walk.first.major, walk.first.ino,
```
I was totally wondering about this and I noticed the failure pattern as well, thanks!

aversecat commented on kmod/src/ioctl.c:
```diff
@@ -113,77 +115,109 @@ static long scoutfs_ioc_walk_inodes(struct file *file, unsigned long arg)
-				goto out;
+			ent->major = le64_to_cpu(key.skii_major);
+			ent->minor = 0;
+			ent->ino = le64_to_cpu(key.skii_ino);
```
It makes sense to me to just zero out the whole page at the start, and then we can omit assigning ent->minor and separately doing the _pad, so I'll just do that.

aversecat commented on kmod/src/data.c:
```diff
@@ -1889,6 +1889,136 @@ int scoutfs_data_waiting(struct super_block *sb, u64 ino, u64 iblock,
+
+	err = __block_write_begin(page, pos, PAGE_SIZE, scoutfs_get_block);
+	if (err) {
+		ret = VM_FAULT_ERROR;
```
Yeahhh, it should be `ret |= VM_FAULT_ERROR;
`

aversecat commented on kmod/src/data.c:
```diff
@@ -1889,6 +1889,136 @@ int scoutfs_data_waiting(struct super_block *sb, u64 ino, u64 iblock,
+	sb_end_pagefault(sb);
+
+	if (scoutfs_data_wait_found(&dw)) {
+		err = scoutfs_data_wait(inode, &dw);
```
Yes, that makes total sense.

aversecat commented:
Next revision below - I've marked all current comments as resolved.

versity-github commented:
Build 203 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/203/ git: f8e216d47075b0ee35fde747e775cc7c155a6466

versity-github commented:
Build 179 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/179/ git: f8e216d47075b0ee35fde747e775cc7c155a6466

zabbo commented on kmod/src/data.c:
```diff
@@ -1912,6 +1912,8 @@ static int scoutfs_data_page_mkwrite(struct vm_area_struct *vma,
 
 	sb_start_pagefault(sb);
 
+	pos = vmf->pgoff << PAGE_SHIFT;
```
It looks like this pos = movement has gotten out of sync with the current version of _data_page_mkwrite.  It now adds two pos = assigments, and this moves one, so we're left with two.  I guess we have _page_mkwrite just add one at the top and remove all the pos motion (and commit message) from this.

aversecat replied:
Yeahh this is a rebase accident...

zabbo commented on kmod/src/ioctl.c:
```diff
@@ -113,77 +115,110 @@ static long scoutfs_ioc_walk_inodes(struct file *file, unsigned long arg)
-				ret = 0;
-				goto out;
-			}
+			nr++;
```
nr is returned if it's positive, even if we're bailing 'cause we saw an error.  We only want to increment it after we've successfully copied to userspace.  (so the nr >= nr_entries test would be (nr + entries)).

zabbo commented on kmod/src/ioctl.c:
```diff
@@ -113,77 +115,110 @@ static long scoutfs_ioc_walk_inodes(struct file *file, unsigned long arg)
-			if (scoutfs_key_compare(&last_key, &lock->end) <= 0) {
-				ret = 0;
+		/* inner loop 1 */
+		while (ent < end) {
```
Annoyingly, we can have a partial entry extend past the end of the page  (page_size % sizeof(ent) != 0).  It turns out to be OK 'cause the part of the entry that extents past the page is never assigned, but that'd be an exciting rule to later break.  Like the -1 or next_aligned tests, we need to test here that storing into the pointer doesn't exceed, not that it starts before the end of the page.  so (ent + 1) <= end would do it, but there's a bunch of ways to express it.

zabbo commented on kmod/src/data.c:
```diff
@@ -1565,48 +1564,93 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
-	}
+		/* still one left, it's in cur */
+		if (cur.len)
+			ret = fill_extent(fieinfo, &cur, last_flags);
```
We don't want to emit the 'cur' extent in the loop that's emitting the buffer of completed extents.  It's not done, it's the extent that we're currently building.  Doing it this way will emit duplicates as first you'll emit it as its merging and then later emit the resulting merged extent that was set in the page.  There's lots of indirect evidence of this, like it not checking the result of ret -- it just fell through to exiting -- or being the only user of last_flags that's only set once we're done.  Anyway, drop this down to be out of the loop body and in the exit path.

zabbo reviewed (changes requested):
More little stuff, feels so close!

aversecat commented on kmod/src/data.c:
```diff
@@ -1889,6 +1889,147 @@ int scoutfs_data_waiting(struct super_block *sb, u64 ino, u64 iblock,
+	wait_for_stable_page(page);
+
+	/* start generic_write_end */
+	//XXX __ocfs2_page_mkwrite has extended error checking around truncate/extend
```
Dropping this comment. Back traced everything and I think it was resolved with the `!PageUptodate..` checks. The comment was wrong because those checks were in _write_begin_nolock() in ocfs2 so this comment was in the wrong spot to begin with.

aversecat commented:
Updated again. I found another incorrect VM_FAULT_ERROR usage and fixed that one up. Fixed up the VM_FAULT_RETRY path in mkwrite causing double locked mmap_sem with VM_FAULT_NOPAGE and triple ran this through tests.

versity-github commented:
Build 212 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/212/ git: fb43a751a4263e33d16d8f496fb7fa9ab33e6521

versity-github commented:
Build 188 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/188/ git: fb43a751a4263e33d16d8f496fb7fa9ab33e6521

aversecat commented:
one more small compat fix for el7.

versity-github commented:
Build 189 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr-el8/189/ git: 9f689e2995c19cf78cffc5bc008f73bb1162cd64

versity-github commented:
Build 213 http://jenkins.vpn.versity.com:8080/job/scoutfs-run-pr/213/ git: 9f689e2995c19cf78cffc5bc008f73bb1162cd64

aversecat commented:
Don't review, this is just a rebase on `el9` and CI check

aversecat commented:
retest

aversecat commented:
the `-debug-` failures are:

 - `ng-scoutfs-test-debug-el7`: 
```
12:56:06   lock-recover-invalidate          [ failed: output differs ]
```

The output has `bash: terminated` interspersed. This is sometimes happening where stderr/stdout for the subprocess isn't properly discarded by redirection to /dev/null.

 - `ng-scoutfs-test-debug-el8`:
```
12:38:41   large-fragmented-free            [ failed: unexpected messages in dmesg ]
16:41:21   orphan-inodes                    [ failed: output differs ]
```

the dmesg here is `blocked for more than 120 seconds.` and subsequent stack dumps.

The `output differs` here is this intermittent test failure pattern:
```
 == orphaned inos in all mounts all deleted
+5264385 still exists
+5274624 still exists
+5284864 still exists
+5295104 still exists
+5305344 still exists
```

 - `ng-scoutfs-test-debug-el94`:
```
12:40:13   large-fragmented-free            [ failed: unexpected messages in dmesg ]
16:05:55   createmany-rename-large-dir      [ failed: unexpected messages in dmesg ]
13:45:10   srch-basic-functionality         [ failed: unexpected messages in dmesg ]
 ```
 
Multiple `watchdog: BUG: soft lockup - CPU#1 stuck for 23s! [kworker/u4:20:18707]` here - the VM is clearly overrun by scheduled tasks.

zabbo commented:
> the `-debug-` failures are:

Thanks for enumerating these -- yeah, so, the same repeat offenders :/.

aversecat commented:
Hmmmm, not looking good in CI.

Both debug-el8 and debug-el94 appear(*) stuck on `generic/346` which is `holetest` with mmap. Trying to reproduce myself now.

(*) they completed `generic/343` but output is stuck there. No message indicating 346 actually started.

aversecat commented:
`-debug-` failures are:

 - `el8`: hung task timeout in `large-fragmented-free`, and `orphan-inodes` failure
 - `el94`: hung task timeout in `large-fragmented-free`, and `orphan-inodes` failure
 - `el95`: hung task timeout in `large-fragmented-free`, and this one:

```
srch-basic-functionality unexpected messages in dmesg
--- golden/archive-light-cycle	2024-12-06 20:05:22.572655167 +0000
+++ /root/scoutfs/tests/results/output/archive-light-cycle	2024-12-07 03:12:08.082647728 +0000
@@ -4,6 +4,8 @@
 == round 1: create
 == round 1: online
 == round 1: verify
+/mnt/test.3/test/archive-light-cycle/dir/3/2-2 /dev/fd/63 differ: char 277401601, line 67726
+script pid 165368 failed: rc 1
 == round 1: release
 == round 1: offline
 == round 1: stage
archive-light-cycle output differs
```

zabbo commented:
retest

zabbo commented on kmod/src/dir.c:
```diff
@@ -447,69 +458,103 @@ static int scoutfs_readdir(struct file *file, struct dir_context *ctx)
 		}
 
-		pos = le64_to_cpu(key.skd_major);
-		ctx->pos = pos;
```
Removing this setting of ctx->pos so the dirente's pos before calling dir_emit is messing up d_off in getdents.  It's a very, very, confusing interface.  The shortest story is that at each call ctx->pos has to be set to the offset of the entry for the call.  (d_off is actually the offset of the next dirent, and filldir sets the *previous* dirent's d_off on a call).  You can see this in how others like dir_emit_dots, btrfs, or bcachefs call.  set ctx->pos, call dir_emit, and if it succeed increment ctx->pos.  That's what the old code did, we just need to maintain that.

aversecat replied:
Found the culprit and I think I have a relatively simple solution below, with tests to verify.

aversecat commented:
@zabbo : Mixed push for easier "re-review":

Added [e59a5f8](https://github.com/versity/scoutfs/pull/175/commits/e59a5f8ebda5b17c3cce3afbeecb76da5f01cc05) Which adds a readdir output test using xfs_io. The golden here was generated using `main`.

Added [e9d1472](https://github.com/versity/scoutfs/pull/175/commits/e9d147260c2796d21b18b111f1e3b62d01ddda21) which fixes the output of `dir_emit` to properly emit using proper `ctx->pos` values for each entry. This then now passes the test.

We could merge as is, or we can let this run through tests once more and squash the fix into @aversecat
[Avoid deadlock in _readdir() due to copy_to_user().](https://github.com/versity/scoutfs/pull/175/commits/cad12d5ce80ea9d4af15b8f10b385dbb6e1358ec)

Either obviously is fine.

zabbo commented:
There seem to be real failures in archive-light-cycle in the 9.5 debug run.  Have we looked into this?  Are we seeing this anywhere else?

 ```@@ -4,6 +4,7 @@
 == round 1: create
 == round 1: online
 == round 1: verify
+/mnt/test.1/test/archive-light-cycle/dir/1/2-1 /dev/fd/63 differ: char 238444545, line 58215
 == round 1: release
 == round 1: offline
 == round 1: stage```

aversecat commented:
@zabbo I found this in a test run in a branch(#202) that has no `mmap` patches:

```
--- golden/archive-light-cycle	2025-02-13 18:55:06.532816042 +0000
+++ /root/scoutfs/tests/results/output/archive-light-cycle	2025-02-14 02:04:50.421626043 +0000
@@ -15,6 +15,7 @@
 == round 2: create
 == round 2: online
 == round 2: verify
+/mnt/test.4/test/archive-light-cycle/dir/4/2-1 /dev/fd/63 differ: char 134172673, line 32758
 == round 2: release
 == round 2: offline
 == round 2: stage
archive-light-cycle output differs
--- golden/block-stale-reads	2025-02-13 18:55:06.541816111 +0000
+++ /root/scoutfs/tests/results/output/block-stale-reads	2025-02-14 02:04:53.510648267 +0000
@@ -1,2 +1,2 @@
 == Issue scoutfs df to force block reads to trigger stale invalidation/retry
-counter block_cache_remove_stale changed
+counter block_cache_remove_stale didn't change
block-stale-reads output differs
```

So it seems this is a potential issue in `main`.

Here's the `fail.log` for that run: http://jenkins.vpn.versity.com:8080/job/ng-scoutfs-test-debug-el95/16/artifact/artifacts/results/fail.log/*view*/

zabbo commented:
> So it seems this is a potential issue in `main`.

So it would seem, and so far only 9.5 debug runs I think.

aversecat commented:
I checked 2 of the offsets, and they're suspiciously at 1byte beyond a 4k boundary:

```
$ printf "0x%08x\n" 203247617
0x0c1d5001
$ printf "0x%08x\n" 203247617
0x0c1d5001
```
